subsys/net/lib/lwm2m/lwm2m_engine.c

```diff
@@ -651,17 +786,163 @@ static int engine_add_observer(struct lwm2m_message *msg,
+	}
+	return list_size;
+}
 
-	if (ctx->observe_cb) {
-		ctx->observe_cb(LWM2M_OBSERVE_EVENT_OBSERVER_REMOVED, &obs->path, NULL);
+static bool engine_compare_obs_path_list(sys_slist_t *obs_path_list, sys_slist_t *path_list,
+					 int list_length)
+{
+	sys_snode_t *obs_ptr, *comp_ptr;
+	struct lwm2m_obj_path_list *obs_path, *comp_path;
+
+	obs_ptr = sys_slist_peek_head(obs_path_list);
+	comp_ptr = sys_slist_peek_head(obs_path_list);
+	while (list_length--) {
+		obs_path = (struct lwm2m_obj_path_list *) obs_ptr;
+		comp_path = (struct lwm2m_obj_path_list *) comp_ptr;
+		if (memcmp(&obs_path->path, &comp_path->path, sizeof(struct lwm2m_obj_path))) {
+			return false;
+		}
+		/* Read Next Info from list entry */
+		obs_ptr = sys_slist_peek_next_no_check(obs_ptr);
+		comp_ptr = sys_slist_peek_next_no_check(comp_ptr);
+	}
+
+	return true;
+}
+
+static int engine_add_composite_observer(struct lwm2m_message *msg,
+			       const uint8_t *token, uint8_t tkl,
+			       uint16_t format)
+{
+	struct observe_node *obs;
+	struct lwm2m_obj_path_list *entry, *tmp;
+	struct notification_attrs attrs;
+	struct lwm2m_obj_path_list lwm2m_path_list_buf[CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE];
+	sys_slist_t lwm_path_list;
+	sys_slist_t lwm_path_free_list;
+	int ret, obs_list_size, path_list_size;
+
+	if (!msg || !msg->ctx) {
+		LOG_ERR("valid lwm2m message is required");
+		return -EINVAL;
+	}
+
+	if (!token || (tkl == 0U || tkl > MAX_TOKEN_LEN)) {
+		LOG_ERR("token(%p) and token length(%u) must be valid.",
+			token, tkl);
+		return -EINVAL;
+	}
+
+	/* Init list */
+	lwm2m_engine_path_list_init(&lwm_path_list, &lwm_path_free_list, lwm2m_path_list_buf,
+				    CONFIG_LWM2M_COMPOSITE_PATH_LIST_SIZE);
+
+	ret = do_composite_observe_read_path_op(msg, format, &lwm_path_list, &lwm_path_free_list);
+	if (ret < 0) {
+		return ret;
+	}
+	path_list_size = engine_path_list_size(&lwm_path_list);
+
+	/* TODO: observe dup checking */
+
+	/* make sure this observer doesn't exist already */
+	SYS_SLIST_FOR_EACH_CONTAINER(&msg->ctx->observer, obs, node) {
+		if (!obs->composite) {
+			continue;
+		}
+		obs_list_size = engine_path_list_size(&obs->path_list);
+
+		if (obs_list_size != path_list_size) {
+			continue;
+		}
+
+		/* TODO: distinguish server object */
+		if (engine_compare_obs_path_list(&obs->path_list, &lwm_path_list, obs_list_size)) {
+			/* quietly update the token information */
+			memcpy(obs->token, token, tkl);
+			obs->tkl = tkl;
+
+			LOG_DBG("OBSERVER Composite DUPLICATE [%s]",
+				log_strdup(
+				lwm2m_sprint_ip_addr(&msg->ctx->remote_addr)));
+
+			return 0;
+		}
+	}
+
+	ret = engine_observe_attribute_list_get(&lwm_path_list, &attrs, msg->ctx->srv_obj_inst);
+	if (ret < 0) {
+		return ret;
+	}
+
+	obs = engine_allocate_observer();
+	if (!obs) {
+		return -ENOMEM;
+	}
+
+	/* Allocate and copy path */
+	SYS_SLIST_FOR_EACH_CONTAINER(&lwm_path_list, tmp, node) {
+		/* Allocate path entry */
+		entry = lwm2m_engine_get_from_list(&obs_obj_path_list);
+
```

Comment:
nit really, but I think it's cleaner to stick the result check (`if (!entry) {`) to the actual operation, and leave an empty line afterward.

Reply:
ok